### ISSUE TITLE
Reactive client wrapping mapped exceptions in ClientWebApplicationException

### ISSUE THREAD

**Issue by lw-mcno** [quarkusio/quarkus#37029]
### Describe the bug

Migrating from rest client classic to reactive, the exception handling behavior is different.

Up until now, a custom `org.eclipse.microprofile.rest.client.ext.ResponseExceptionMapper` was in use (enabled using the resources `org.eclipse.microprofile.rest.client.spi.RestClientBuilderListener` SPI-listing). It is outputting instances of `jakarta.ws.rs.ClientErrorException`, e.g. the `jakarta.ws.rs.NotFoundException`, depending on the HTTP response status code.

This allowed for easy catching and servicing of exceptions based on concrete result type, e.g.
```java
try {
    doSomething();
} catch (NotFoundException xxx) {
    reactToNotFound();
} ...
```

The reactive client is wrapping all instances of `jakarta.ws.rs.WebApplicationException` (that is, also the `ClientErrorException`) in its own `org.jboss.resteasy.reactive.ClientWebApplicationException`, where the original mapped exception is the cause. This requires catching the general exception (`org.jboss.resteasy.reactive.ClientWebApplicationException` or `jakarta.ws.rs.WebApplicationException`) and only then looking at what it really contains, which doesn't allow for writing separate, specific catch clauses.

This happens in `org.jboss.resteasy.reactive.client.impl.RestClientRequestContext#unwrapException`.

What is the general / cohesive idea for correct (and preferably simple) exception handling here? The wrapping is not mandated by the spec (3.0 at the moment) and maybe even acts against the intention of response exception mappers.

### Expected behavior

The exceptions are thrown by clients as returned by response exception mapper(s).

### Actual behavior

The mapped exceptions are additionally wrapped.

### How to Reproduce?

Create and register (as SPI using a file in resources/), the following MicroProfile client exception mapper:

```java
public class ClientResponseExceptionMapper implements ResponseExceptionMapper<WebApplicationException> {

    @Override
    public WebApplicationException toThrowable(Response response) {

        Response.Status status = response.getStatusInfo().toEnum();

        if (status == null) {
            return null;
        }

        switch (status) {
            case NOT_FOUND:
                return new NotFoundException(response); // in reality a custom type extending the exception gets returned, not included here for simplicity
            default:
                return new ClientErrorException(response);
        }

        return null;
    }
```

Create a client and call an endpoint returning 404. Observe the returned exception type and its cause.

**Comment by lw-mcno** [comment]
Info from @Ladicek: this (the wrapping) is done supposedly because otherwise service exception mappers could treat such an uncaught `WebApplicationException` as the service's own exception, possibly leaking some upstream server's response data to the caller. One method of working around this is to use a set of exceptions that are not WebApplicationException-s.

There's still the point of behavioral difference between REST client classic and reactive.

**Comment by cescoffier** [comment]
Yes, but I think the reactive client has the right behavior in this case.

**Comment by nderwin** [comment]
Unfortunately, wrapping all exceptions into a single wrapper makes the reactive clients harder to use with the `@Retry` annotation from the fault tolerance extension - you lose the ability to retry on some errors (say 5xx) and not on others (say 4xx) because they are all wrapped into the same exception.

**Comment by geoand** [comment]
> Info from @Ladicek: this (the wrapping) is done supposedly because otherwise, the (server) exception mappers could treat such an uncaught `WebApplicationException` as the service's own exception, possibly leaking some upstream server's response data to the caller. One method of working around this is to use a set of exceptions that are not WebApplicationException-s.
> 
> There's still the point of behavioral difference between REST client classic and reactive.

I am pretty sure that @FroMage has encountered that problem many times in the past

**Comment by geoand** [comment]
> Yes, but I think the reactive client has the right behavior in this case.

Agreed

**Comment by geoand** [comment]
> Unfortunately, wrapping all exceptions into a single wrapper makes the reactive clients harder to use with the `@Retry` annotation from the fault tolerance extension - you lose the ability to retry on some errors (say 5xx) and not on others (say 4xx) because they are all wrapped into the same exception.

@Ladicek would it not be possible to add support for the `cause` of an exception in Smallrye Fault Tolerance?

**Comment by Ladicek** [comment]
SmallRye Fault Tolerance already supports inspecting exception causes in case the original exception doesn't match. This only works in the non-compatible mode, but that is on by default in Quarkus.

It is unfortunately pretty counterintuitive when the configuration of `@Retry` (or `@Fallback` or others) uses the [default] values of `Exception` or `Throwable`, because then, the original exception always matches and we don't get to inspect the cause.

**Comment by FroMage** [comment]
Yes, this was a big CVE in the past, where client exceptions were taken at face-value as if they were server exceptions. Wrapping them is the safest way to make sure this cannot happen, even when non-framework bits (like your `ResponseExceptionMapper`) throw them.

Unfortunately, we don't control the class hierarchies so we can't add a `boolean isClient()` property to `WebApplicationException` to allow clients to use them. Now, if you want to throw custom subtypes of `ClientWebApplicationException` from your `ResponseExceptionMapper`, to indicate special behaviour, I _assume_ those would not be wrapped by the client, allowing you to `catch` them easily, no?

Something like `ClientNotFoundException extends ClientWebApplicationException`?

**Comment by lw-mcno** [comment]
From this piece of code, it looks like subclasses of `ClientWebApplicationException` will get their cause wrapped:

```java
        if (res instanceof WebApplicationException) {
            var webApplicationException = (WebApplicationException) res;
            ...
            return new ClientWebApplicationException(message,
                    webApplicationException instanceof ClientWebApplicationException ? webApplicationException.getCause()
                            : webApplicationException,
                    webApplicationException.getResponse());
        }
```

I'd say if they're already `ClientWebApplicationException`, they shouldn't be repacked at all?

As a cleaner solution, I completely detached my exceptions from this class hierarchy.

**Comment by AndreasPetersen** [comment]
In Quarkus 3.15.1, exceptions from a `ResponseExceptionMapper` are no longer wrapped in a `ClientWebApplicationException`. This makes the exceptions leak to the server. From my understanding of this issue, that is not the intended behaviour?

In Quarkus 3.12.1 the exceptions were wrapped in a `ClientWebApplicationException` as described in this issue. I couldn't find anything in the 3.13, 3.14 or 3.15 migration guides about this change.

Example of a `ResponseExceptionMapper`:

```
public class NotFoundExceptionMapper
        implements ResponseExceptionMapper<WebApplicationException> {

    @Override
    public boolean handles(int status, MultivaluedMap<String, Object> headers) {
        return status == HttpURLConnection.HTTP_NOT_FOUND;
    }

    @Override
    public WebApplicationException toThrowable(Response response) {
        return new NotFoundException(response);
    }
}
```

**Comment by geoand** [comment]
@AndreasPetersen can you please attach a sample that behaves like you say so we can see the difference between 3.12 and 3.15 in action?

Thanks

**Comment by geoand** [comment]
Thanks for the reproducer!

The change was introduced in https://github.com/quarkusio/quarkus/issues/42275

**Comment by FroMage** [comment]
See https://github.com/quarkusio/quarkus/pull/42290#issuecomment-2265324908
